File: lib/investments.ts
import { and, asc, desc, eq, inArray, isNotNull } from "drizzle-orm";
import { db } from "@/lib/db";
import {
  accountGroups,
  accounts,
  portfolioValuations,
  transactions,
  type Account,
  type AccountGroup,
  type PortfolioValuation,
} from "@/db/schema";
import { accountInSpaceClause } from "@/lib/spaces";

export interface InvestmentAccountRow {
  account: Account;
  group: AccountGroup;
}

// Intentionally ignores accounts.excluded: investment accounts are typically
// excluded from cash net worth, but they must still appear on /investments.
export async function listInvestmentAccountsInSpace(
  spaceId: string,
  defaultSpaceId: string,
): Promise<InvestmentAccountRow[]> {
  return db
    .select({ account: accounts, group: accountGroups })
    .from(accounts)
    .innerJoin(accountGroups, eq(accounts.groupId, accountGroups.id))
    .where(
      and(
        accountInSpaceClause(spaceId, defaultSpaceId),
        eq(accounts.archived, false),
        eq(accountGroups.kind, "investment"),
      ),
    );
}

export async function listValuationsForAccounts(
  accountIds: string[],
): Promise<PortfolioValuation[]> {
  if (accountIds.length === 0) return [];
  return db
    .select()
    .from(portfolioValuations)
    .where(inArray(portfolioValuations.accountId, accountIds))
    .orderBy(asc(portfolioValuations.asOf));
}

export interface NetContributionLeg {
  accountId: string;
  bookedAt: Date;
  netEur: number;
}

/**
 * All transfer legs touching the given investment accounts. Each leg counts
 * toward its own account's cost basis, signed: credits positive, debits negative.
 *
 * Inter-investment moves (Investment → Pension, Trading212 → MyInvestor) are
 * NOT excluded: each leg shifts its own account's cost basis, and the totals
 * across accounts naturally cancel (the debit on one is the credit on the
 * other), so the portfolio-wide cost basis stays correct while per-account
 * numbers reflect the reallocation.
 */
export async function listInvestmentContributionLegs(
  investmentAccountIds: string[],
): Promise<NetContributionLeg[]> {
  if (investmentAccountIds.length === 0) return [];
  const rows = await db
    .select({
      accountId: transactions.accountId,
      bookedAt: transactions.bookedAt,
      amountEur: transactions.amountEur,
    })
    .from(transactions)
    .where(
      and(
        inArray(transactions.accountId, investmentAccountIds),
        eq(transactions.isTransfer, true),
        isNotNull(transactions.transferGroupId),
        isNotNull(transactions.amountEur),
      ),
    )
    .orderBy(asc(transactions.bookedAt));

  return rows.map((r) => ({
    accountId: r.accountId,
    bookedAt: r.bookedAt,
    netEur: Number(r.amountEur),
  }));
}

export interface AccountMetrics {
  accountId: string;
  baselineAsOf: Date | null;
  baselineEur: number | null;
  latestAsOf: Date | null;
  latestEur: number | null;
  latestCashEur: number | null;
  latestPositionsEur: number | null;
  netContributionsSinceBaselineEur: number;
  costBasisEur: number | null;
  pnlEur: number | null;
  pnlPct: number | null;
}

function emptyMetrics(accountId: string): AccountMetrics {
  return {
    accountId,
    baselineAsOf: null,
    baselineEur: null,
    latestAsOf: null,
    latestEur: null,
    latestCashEur: null,
    latestPositionsEur: null,
    netContributionsSinceBaselineEur: 0,
    costBasisEur: null,
    pnlEur: null,
    pnlPct: null,
  };
}

/**
 * Per-account metrics: baseline (earliest valuation), current value (latest valuation),
 * cost basis (baseline + net contributions after baseline date), P&L.
 */
export function computeAccountMetrics(
  investmentAccountIds: string[],
  valuations: PortfolioValuation[],
  legs: NetContributionLeg[],
): Map<string, AccountMetrics> {
  const valsByAccount = new Map<string, PortfolioValuation[]>();
  for (const v of valuations) {
    const arr = valsByAccount.get(v.accountId) ?? [];
    arr.push(v);
    valsByAccount.set(v.accountId, arr);
  }

  const out = new Map<string, AccountMetrics>();
  for (const accId of investmentAccountIds) {
    const list = valsByAccount.get(accId) ?? [];
    if (list.length === 0) {
      out.set(accId, emptyMetrics(accId));
      continue;
    }
    const baseline = list[0];
    const latest = list.at(-1)!;
    const baselineTime = baseline.asOf.getTime();
    let net = 0;
    for (const leg of legs) {
      if (leg.accountId !== accId) continue;
      // Strict >: the baseline already reflects everything as of its date, so
      // same-day moves are NOT added on top. Otherwise a snapshot that
      // implicitly includes today's deposit would double-count it.
      if (leg.bookedAt.getTime() > baselineTime) net += leg.netEur;
    }
    const baselineEur = Number(baseline.marketValueEur);
    const latestEur = Number(latest.marketValueEur);
    const latestCashEur = latest.cashValueEur != null ? Number(latest.cashValueEur) : null;
    const latestPositionsEur =
      latestCashEur != null ? Math.max(0, latestEur - latestCashEur) : null;
    const costBasis = baselineEur + net;
    const pnl = latestEur - costBasis;
    const pnlPct = Math.abs(costBasis) > 1e-6 ? pnl / costBasis : null;
    out.set(accId, {
      accountId: accId,
      baselineAsOf: baseline.asOf,
      baselineEur,
      latestAsOf: latest.asOf,
      latestEur,
      latestCashEur,
      latestPositionsEur,
      netContributionsSinceBaselineEur: net,
      costBasisEur: costBasis,
      pnlEur: pnl,
      pnlPct,
    });
  }
  return out;
}

export interface PortfolioSeriesPoint {
  date: string; // YYYY-MM-DD
  marketValueEur: number;
  costBasisEur: number;
  cashEur: number;
  positionsEur: number;
}

/**
 * Time series with one point per distinct valuation date across all accounts.
 * Per account at date d:
 *   marketValue = latest valuation with asOf <= d (carry-forward)
 *   costBasis   = baselineEur + sum(legs.netEur where bookedAt in (baselineDate, d])
 * Sums across accounts.
 */
export function computePortfolioSeries(
  investmentAccountIds: string[],
  valuations: PortfolioValuation[],
  legs: NetContributionLeg[],
): PortfolioSeriesPoint[] {
  if (valuations.length === 0) return [];

  const valsByAccount = new Map<string, PortfolioValuation[]>();
  for (const v of valuations) {
    const arr = valsByAccount.get(v.accountId) ?? [];
    arr.push(v);
    valsByAccount.set(v.accountId, arr);
  }

  const dateSet = new Set<string>();
  for (const v of valuations) dateSet.add(v.asOf.toISOString().slice(0, 10));
  const dates = Array.from(dateSet).sort((a, b) => a.localeCompare(b));

  const series: PortfolioSeriesPoint[] = [];
  for (const d of dates) {
    const endTime = new Date(d + "T23:59:59.999Z").getTime();
    let marketValue = 0;
    let cashTotal = 0;
    let costBasis = 0;
    for (const accId of investmentAccountIds) {
      const list = valsByAccount.get(accId);
      if (!list || list.length === 0) continue;
      const baseline = list[0];
      const baselineTime = baseline.asOf.getTime();
      if (baselineTime > endTime) continue;
      let mv = 0;
      let cash = 0;
      for (const v of list) {
        if (v.asOf.getTime() <= endTime) {
          mv = Number(v.marketValueEur);
          cash = v.cashValueEur != null ? Number(v.cashValueEur) : 0;
        } else {
          break;
        }
      }
      marketValue += mv;
      cashTotal += cash;
      let contrib = 0;
      for (const leg of legs) {
        if (leg.accountId !== accId) continue;
        const t = leg.bookedAt.getTime();
        if (t > baselineTime && t <= endTime) contrib += leg.netEur;
      }
      costBasis += Number(baseline.marketValueEur) + contrib;
    }
    const positions = Math.max(0, marketValue - cashTotal);
    series.push({
      date: d,
      marketValueEur: marketValue,
      costBasisEur: costBasis,
      cashEur: cashTotal,
      positionsEur: positions,
    });
  }

  return series;
}

/**
 * Sum of latest valuations across the given investment accounts (EUR).
 * Used by the dashboard to compute Total Net Worth = cash + investments.
 */
export async function sumLatestInvestmentValueEur(
  investmentAccountIds: string[],
): Promise<number> {
  if (investmentAccountIds.length === 0) return 0;
  const rows = await db
    .select({
      accountId: portfolioValuations.accountId,
      marketValueEur: portfolioValuations.marketValueEur,
    })
    .from(portfolioValuations)
    .where(inArray(portfolioValuations.accountId, investmentAccountIds))
    .orderBy(desc(portfolioValuations.asOf));
  const seen = new Set<string>();
  let total = 0;
  for (const r of rows) {
    if (seen.has(r.accountId)) continue;
    seen.add(r.accountId);
    total += Number(r.marketValueEur);
  }
  return total;
}

/**
 * Account ids in the space whose group has kind='investment'. Used to feed the
 * dashboard's Total Net Worth computation.
 */
export async function listInvestmentAccountIdsInSpace(
  spaceId: string,
  defaultSpaceId: string,
): Promise<string[]> {
  const rows = await db
    .select({ id: accounts.id })
    .from(accounts)
    .innerJoin(accountGroups, eq(accounts.groupId, accountGroups.id))
    .where(
      and(
        accountInSpaceClause(spaceId, defaultSpaceId),
        eq(accounts.archived, false),
        eq(accountGroups.kind, "investment"),
      ),
    );
  return rows.map((r) => r.id);
}

export type InvestmentPeriod = "ytd" | "1y" | "3y" | "all";

export function periodStartDate(period: InvestmentPeriod, now = new Date()): Date | null {
  if (period === "all") return null;
  if (period === "ytd") return new Date(Date.UTC(now.getUTCFullYear(), 0, 1));
  if (period === "1y") {
    const d = new Date(now);
    d.setUTCFullYear(d.getUTCFullYear() - 1);
    return d;
  }
  const d = new Date(now);
  d.setUTCFullYear(d.getUTCFullYear() - 3);
  return d;
}

File: Core/Sources/CoreLogic/Investments.swift
import Foundation
import SwiftData
import CoreModel

extension CoreLogic {
    public enum Investments {
        public enum Period: String, CaseIterable, Sendable {
            case ytd
            case oneYear = "1y"
            case threeYears = "3y"
            case all
        }

        public struct InvestmentAccountRow {
            public let account: Account
            public let group: AccountGroup
        }

        public struct ContributionLeg: Equatable, Sendable {
            public let accountId: UUID
            public let bookedAt: Date
            public let netEur: Decimal
        }

        // Both sides of an investment account are an anchor plus every leg booked after it —
        // the same shape as Account.balanceAnchor for cash. Cost basis anchors on
        // costBasisOpeningEur, market value on the newest snapshot. A deposit therefore lifts
        // value and cost by the identical amount and leaves P&L alone, which is the whole
        // point: paying money in is not a loss.
        public struct AccountMetrics: Equatable, Sendable {
            public let accountId: UUID
            public let latestAsOf: Date?
            // Snapshot (or live) figure, before contributions booked after it.
            public let anchorValueEur: Decimal?
            // anchorValueEur + contributions since. What the UI shows.
            public let valueEur: Decimal?
            public let latestCashEur: Decimal?
            public let latestPositionsEur: Decimal?
            // Paid in since the value anchor — the part the snapshot can't know about.
            public let contributionsSinceValueEur: Decimal
            public let costBasisEur: Decimal?
            public let pnlEur: Decimal?
            public let pnlPct: Decimal?
            // True when the value came from a live feed rather than a typed snapshot.
            public let isLive: Bool

            public var isStale: Bool {
                guard !isLive, let latestAsOf else { return false }
                return Date.now.timeIntervalSince(latestAsOf) > 45 * 86_400
            }
        }

        public struct PortfolioSeriesPoint: Equatable, Sendable {
            public let date: Date           // UTC start-of-day
            public let marketValueEur: Decimal
            public let costBasisEur: Decimal
            public let cashEur: Decimal
            public let positionsEur: Decimal
        }

        // MARK: - DB readers

        @MainActor
        public static func listAccountsInSpace(
            spaceId: UUID,
            defaultSpaceId: UUID,
            in ctx: ModelContext
        ) throws -> [InvestmentAccountRow] {
            let accounts = try fetchAccountsInSpace(
                spaceId: spaceId, defaultSpaceId: defaultSpaceId, in: ctx
            )
            return accounts.compactMap { acc in
                guard let g = acc.group, g.kind == .investment else { return nil }
                return InvestmentAccountRow(account: acc, group: g)
            }
        }

        @MainActor
        public static func listAccountIdsInSpace(
            spaceId: UUID,
            defaultSpaceId: UUID,
            in ctx: ModelContext
        ) throws -> [UUID] {
            try listAccountsInSpace(
                spaceId: spaceId, defaultSpaceId: defaultSpaceId, in: ctx
            ).map { $0.account.id }
        }

        @MainActor
        public static func listValuations(
            for accountIds: [UUID],
            in ctx: ModelContext
        ) throws -> [PortfolioValuation] {
            if accountIds.isEmpty { return [] }
            let idSet = Set(accountIds)
            let all = try ctx.fetch(FetchDescriptor<PortfolioValuation>(
                predicate: #Predicate { $0.account != nil },
                sortBy: [SortDescriptor(\.asOf, order: .forward)]
            ))
            return all.filter { v in
                guard let id = v.account?.id else { return false }
                return idSet.contains(id)
            }
        }

        @MainActor
        public static func listContributionLegs(
            for accountIds: [UUID],
            in ctx: ModelContext
        ) throws -> [ContributionLeg] {
            if accountIds.isEmpty { return [] }
            let idSet = Set(accountIds)
            // No transferGroup requirement: an unpaired leg is still money that left one
            // account and landed here. Requiring the pair silently dropped contributions.
            // isTransfer stays, so interest/dividends booked on the account read as return,
            // not as new money paid in.
            let candidates = try ctx.fetch(FetchDescriptor<Transaction>(
                predicate: #Predicate {
                    $0.isTransfer == true && $0.amountEur != nil
                },
                sortBy: [SortDescriptor(\.bookedAt, order: .forward)]
            ))
            return candidates.compactMap { tx in
                guard let accId = tx.account?.id, idSet.contains(accId) else { return nil }
                guard let eur = tx.amountEur else { return nil }
                return ContributionLeg(accountId: accId, bookedAt: tx.bookedAt, netEur: eur)
            }
        }

        public static func basis(for account: Account) -> AccountBasis {
            AccountBasis(
                accountId: account.id,
                costBasisOpeningEur: account.costBasisOpeningEur,
                costBasisOpeningAt: account.costBasisOpeningAt,
                isLiveValued: account.liveValueSource?.isEmpty == false
            )
        }

        @MainActor
        public static func loadMetrics(
            for accounts: [Account],
            in ctx: ModelContext
        ) throws -> [UUID: AccountMetrics] {
            let ids = accounts.map(\.id)
            return computeAccountMetrics(
                bases: accounts.map(basis(for:)),
                valuations: try listValuations(for: ids, in: ctx),
                legs: try listContributionLegs(for: ids, in: ctx)
            )
        }

        // Net worth's investment slice. Uses the same snapshot + contributions-since figure the
        // Investments tab shows, so the two can't disagree.
        @MainActor
        public static func sumLatestValue(
            for accounts: [Account],
            in ctx: ModelContext
        ) throws -> Decimal {
            if accounts.isEmpty { return 0 }
            return try loadMetrics(for: accounts, in: ctx)
                .values.reduce(Decimal(0)) { $0 + ($1.valueEur ?? 0) }
        }

        // MARK: - Pure compute

        // The per-account inputs computeAccountMetrics needs from Account, so the maths stays
        // pure and testable without a ModelContext.
        public struct AccountBasis: Equatable, Sendable {
            public let accountId: UUID
            public let costBasisOpeningEur: Decimal?
            public let costBasisOpeningAt: Date?
            public let isLiveValued: Bool

            public init(
                accountId: UUID,
                costBasisOpeningEur: Decimal? = nil,
                costBasisOpeningAt: Date? = nil,
                isLiveValued: Bool = false
            ) {
                self.accountId = accountId
                self.costBasisOpeningEur = costBasisOpeningEur
                self.costBasisOpeningAt = costBasisOpeningAt
                self.isLiveValued = isLiveValued
            }
        }

        public static func computeAccountMetrics(
            bases: [AccountBasis],
            valuations: [PortfolioValuation],
            legs: [ContributionLeg]
        ) -> [UUID: AccountMetrics] {
            var byAccount: [UUID: [PortfolioValuation]] = [:]
            for v in valuations {
                guard let id = v.account?.id else { continue }
                byAccount[id, default: []].append(v)
            }

            var out: [UUID: AccountMetrics] = [:]
            for basis in bases {
                let accId = basis.accountId
                let list = byAccount[accId] ?? []
                let latest = list.last

                // Cost basis: opening figure plus every leg after it. With no opening date the
                // account's whole ledger counts, which is right for one opened inside the data.
                let costBasis: Decimal? = basis.costBasisOpeningEur.map { opening in
                    let since = basis.costBasisOpeningAt
                    return legs.reduce(opening) { acc, leg in
                        guard leg.accountId == accId else { return acc }
                        if let since, leg.bookedAt <= since { return acc }
                        return acc + leg.netEur
                    }
                }

                guard let latest else {
                    out[accId] = AccountMetrics(
                        accountId: accId, latestAsOf: nil,
                        anchorValueEur: nil, valueEur: nil,
                        latestCashEur: nil, latestPositionsEur: nil,
                        contributionsSinceValueEur: 0,
                        costBasisEur: costBasis, pnlEur: nil, pnlPct: nil,
                        isLive: basis.isLiveValued
                    )
                    continue
                }

                // Strict >: the snapshot already reflects same-instant moves.
                // A live-valued account is current by definition — adding legs would
                // double-count money the feed already sees.
                var sinceValue: Decimal = 0
                if !basis.isLiveValued {
                    for leg in legs where leg.accountId == accId && leg.bookedAt > latest.asOf {
                        sinceValue += leg.netEur
                    }
                }

                let anchor = latest.marketValueEur
                let value = anchor + sinceValue
                let latestCash = latest.cashValueEur
                let latestPositions: Decimal? = latestCash.map { max(0, value - $0) }
                let pnl: Decimal? = costBasis.map { value - $0 }
                let epsilon = Decimal(string: "0.000001")!
                let pnlPct: Decimal? = {
                    guard let pnl, let costBasis, abs(costBasis) > epsilon else { return nil }
                    return pnl / costBasis
                }()

                out[accId] = AccountMetrics(
                    accountId: accId,
                    latestAsOf: latest.asOf,
                    anchorValueEur: anchor,
                    valueEur: value,
                    latestCashEur: latestCash,
                    latestPositionsEur: latestPositions,
                    contributionsSinceValueEur: sinceValue,
                    costBasisEur: costBasis,
                    pnlEur: pnl,
                    pnlPct: pnlPct,
                    isLive: basis.isLiveValued
                )
            }
            return out
        }

        public static func computePortfolioSeries(
            bases: [AccountBasis],
            valuations: [PortfolioValuation],
            legs: [ContributionLeg]
        ) -> [PortfolioSeriesPoint] {
            if valuations.isEmpty { return [] }

            var cal = Calendar(identifier: .iso8601)
            cal.timeZone = TimeZone(identifier: "UTC")!

            var byAccount: [UUID: [PortfolioValuation]] = [:]
            for v in valuations {
                guard let id = v.account?.id else { continue }
                byAccount[id, default: []].append(v)
            }

            // Deposit days are plotted too, so paying money in visibly steps the line rather
            // than staying flat until the next snapshot.
            let firstSnapshot = valuations.map(\.asOf).min()!
            var dateSet = Set(valuations.map { cal.startOfDay(for: $0.asOf) })
            for leg in legs where leg.bookedAt > firstSnapshot {
                dateSet.insert(cal.startOfDay(for: leg.bookedAt))
            }
            let dates = dateSet.sorted()

            var out: [PortfolioSeriesPoint] = []
            for d in dates {
                let endOfDay = cal.date(byAdding: .day, value: 1, to: d)!
                var marketValue: Decimal = 0
                var cashTotal: Decimal = 0
                var costBasisTotal: Decimal = 0
                for basis in bases {
                    let accId = basis.accountId
                    guard let list = byAccount[accId], let first = list.first,
                          first.asOf < endOfDay else { continue }
                    var anchor: Decimal = 0
                    var anchorAt = first.asOf
                    var cash: Decimal = 0
                    for v in list {
                        if v.asOf < endOfDay {
                            anchor = v.marketValueEur
                            anchorAt = v.asOf
                            cash = v.cashValueEur ?? 0
                        } else { break }
                    }
                    var sinceValue: Decimal = 0
                    if !basis.isLiveValued {
                        for leg in legs where leg.accountId == accId
                            && leg.bookedAt > anchorAt && leg.bookedAt < endOfDay {
                            sinceValue += leg.netEur
                        }
                    }
                    marketValue += anchor + sinceValue
                    cashTotal += cash

                    if let opening = basis.costBasisOpeningEur {
                        var contrib: Decimal = 0
                        for leg in legs where leg.accountId == accId && leg.bookedAt < endOfDay {
                            if let since = basis.costBasisOpeningAt, leg.bookedAt <= since {
                                continue
                            }
                            contrib += leg.netEur
                        }
                        costBasisTotal += opening + contrib
                    }
                }
                let positions = max(0, marketValue - cashTotal)
                out.append(PortfolioSeriesPoint(
                    date: d,
                    marketValueEur: marketValue,
                    costBasisEur: costBasisTotal,
                    cashEur: cashTotal,
                    positionsEur: positions
                ))
            }
            return out
        }

        // MARK: - Helpers


        // Records "what this account is worth today". Snapshots are the only investment
        // data the app keeps (Instrument/Holding/Price were deliberately not ported), so
        // without a way to add one the displayed value silently ages — contributions after
        // the last snapshot never show up.
        // Same-day re-entry overwrites rather than stacking, so correcting a typo doesn't
        // leave two snapshots fighting over the same point on the chart.
        @MainActor
        @discardableResult
        public static func recordValuation(
            account: Account,
            marketValueEur: Decimal,
            cashValueEur: Decimal? = nil,
            asOf: Date = .now,
            notes: String? = nil,
            in ctx: ModelContext,
            now: Date = .now
        ) throws -> PortfolioValuation {
            let accountId = account.id
            let day = dayStart(asOf)
            let nextDay = day.addingTimeInterval(86_400)
            let existing = try ctx.fetch(FetchDescriptor<PortfolioValuation>(
                predicate: #Predicate {
                    $0.account?.id == accountId && $0.asOf >= day && $0.asOf < nextDay
                }
            )).first

            if let existing {
                existing.marketValueEur = marketValueEur
                existing.cashValueEur = cashValueEur
                existing.notes = notes
                existing.asOf = asOf
                existing.updatedAt = now
                try ctx.saveTouchingChanges()
                return existing
            }

            let valuation = PortfolioValuation(
                account: account,
                asOf: asOf,
                marketValueEur: marketValueEur,
                cashValueEur: cashValueEur,
                notes: notes,
                createdAt: now
            )
            ctx.insert(valuation)
            try ctx.saveTouchingChanges()
            return valuation
        }


        // The money-in anchor. Everything transferred in after `at` is added automatically,
        // so this is only ever set once, for whatever was already in the account before the
        // ledger starts.
        @MainActor
        public static func setCostBasisOpening(
            _ account: Account, amountEur: Decimal?, at date: Date?, in ctx: ModelContext
        ) throws {
            account.costBasisOpeningEur = amountEur
            account.costBasisOpeningAt = amountEur == nil ? nil : date
            try ctx.saveTouchingChanges()
        }

        @MainActor
        public static func setLiveSource(
            _ account: Account, source: String?, quantity: Decimal?, in ctx: ModelContext
        ) throws {
            let clean = source?.trimmingCharacters(in: .whitespacesAndNewlines)
            account.liveValueSource = (clean?.isEmpty ?? true) ? nil : clean
            account.assetQuantity = account.liveValueSource == nil ? nil : quantity
            try ctx.saveTouchingChanges()
        }

        // Safe now that cost basis lives on the account: deleting a snapshot only drops a
        // market-value reading, it can't destroy the P&L baseline the way it used to.
        @MainActor
        public static func deleteValuation(_ valuation: PortfolioValuation, in ctx: ModelContext) throws {
            ctx.delete(valuation)
            try ctx.saveTouchingChanges()
        }

        private static func dayStart(_ date: Date) -> Date {
            var cal = Calendar(identifier: .iso8601)
            cal.timeZone = TimeZone(identifier: "UTC")!
            return cal.startOfDay(for: date)
        }

        public static func periodStartDate(_ period: Period, now: Date = .now) -> Date? {
            var cal = Calendar(identifier: .iso8601)
            cal.timeZone = TimeZone(identifier: "UTC")!
            switch period {
            case .all:
                return nil
            case .ytd:
                let year = cal.component(.year, from: now)
                return cal.date(from: DateComponents(year: year, month: 1, day: 1))
            case .oneYear:
                return cal.date(byAdding: .year, value: -1, to: now)
            case .threeYears:
                return cal.date(byAdding: .year, value: -3, to: now)
            }
        }

        // The default space catches all accounts with no space assignment.
        // Match TS `accountInSpaceClause` semantics.
        @MainActor
        private static func fetchAccountsInSpace(
            spaceId: UUID,
            defaultSpaceId: UUID,
            in ctx: ModelContext
        ) throws -> [Account] {
            if spaceId == defaultSpaceId {
                return try ctx.fetch(FetchDescriptor<Account>(
                    predicate: #Predicate {
                        $0.archived == false && ($0.space == nil || $0.space?.id == spaceId)
                    }
                ))
            }
            return try ctx.fetch(FetchDescriptor<Account>(
                predicate: #Predicate { $0.archived == false && $0.space?.id == spaceId }
            ))
        }
    }
}
